docs/architecture/psa-migration/md-cipher-dispatch.md

```diff
@@ -263,12 +319,70 @@ These problems are easily solvable.
 * We can make names and HMAC optional. The mixed-domain hash interface won't be the full `MBEDTLS_MD_C` but a subset.
 * We can optimize `md.c` without making API changes to `md.h`.
 
+### Scope reductions and priorities for 3.x
+
+This section documents things that we chose to temporarily exclude from the scope in the 3.x branch (which will eventually be in scope again after 4.0) as well as things we chose to prioritize if we don't have time to support everything.
+
+#### Don't support PK, X.509 and TLS without `MBEDTLS_USE_PSA_CRYPTO`
+
+We do not need to support driver-only hashes and ciphers in PK. X.509 and TLS without `MBEDTLS_USE_PSA_CRYPTO`. Users who want to take full advantage of drivers will need to enabled this macro.
+
+Note that this applies to TLS 1.3 as well, as some uses of hashes and all uses of ciphers there are common with TLS 1.2, hence governed by `MBEDTLS_USE_PSA_CRYPTO`, see [this macro's extended documentation](../../docs/use-psa-crypto.html).
+
+This will go away naturally in 4.0 when this macros is not longer an option (because it's always on).
+
+#### Don't support for `MBEDTLS_PSA_CRYPTO_CLIENT` without `MBEDTLS_PSA_CRYPTO_C`
+
+We generally don't really support builds with `MBEDTLS_PSA_CRYPTO_CLIENT` without `MBEDTLS_PSA_CRYPTO_C`. For example, both `MBEDTLS_USE_PSA_CRYPTO` and `MBEDTLS_SSL_PROTO_TLS1_3` require `MBEDTLS_PSA_CRYPTO_C`, while in principle they should only require `MBEDTLS_PSA_CRYPTO_CLIENT`.
+
+Considering this existing restriction which we do not plan to lift before 4.0, it is acceptable driver-only hashes and cipher support to have the same restriction in 3.x.
+
+It is however desirable for the design to keep support for `MBEDTLS_PSA_CRYPTO_CLIENT` in mind, in order to avoid making it more difficult to add in the future.
+
+#### For cipher: prioritize constrained devices and modern TLS
+
+The primary target is a configuration like TF-M's medium profile, plus TLS with only AEAD ciphersuites.
+
+This excludes things like:
+- Support for encrypted PEM, PKCS5 and PKCS12 encryption, and PKCS8 encrypted keys in PK parse. (Not widely used on highly constrained devices.)
+- Support for NIST-KW. (Same justification.)
+- Support for CMAC. (Same justification, plus can be directly accelerated.)
```

Comment:
The TF-M profile we're currently using excludes CMAC, but [TF-M is pushing the use of CMAC](https://github.com/ARM-software/psa-api/issues/106). That's not CMAC itself, it's a key derivation based on CMAC, but that key derivation is not yet implemented or even fully designed, so in the meantime TF-M users are doing the key derivation computation using the PSA MAC API.

As a consequence, I think it would be good to make CMAC in scope for cipher light 3.6.

I don't think this fundamentally changes the design, but it means more implementation work.

Reply:
Having checked — `PSA_ALG_CMAC` is not important to TF-M. Only the future `PSA_ALG_CMAC_BASED_KDF` is. So, for TF-M, the ideal scenario is if 3.6 supports `PSA_ALG_CMAC_BASED_KDF`, with the smallest code size given what accelerators can do (AES-CMAC, AES block, or none). If 3.6 doesn't support `PSA_ALG_CMAC_BASED_KDF` then it doesn't matter how Mbed TLS's CMAC works: users will go through proprietary APIs to implement a CMAC-based KDF.

It's still unclear at this point whether Mbed TLS 3.6 will support the CMAC-based KDF. It wasn't in our original plan for 3.6, but a few days ago we decided to push the release date of 3.6 back, and this is one feature that might get added to the scope.

Reply:
CMAC is a bit different in that is more tied to the Cipher API:
- Its streaming API works on a `mbedtls_cipher_context_t` object, that the user is supposed to manage using the Cipher API, most notably `mbedtls_cipher_setup()`.
- Its one-shot API takes a `mbedtls_cipher_info_t *` as input.

So, it we want to optimize size for built-in CMAC by making it depend on something smaller than the full Cipher, then that smaller thing pretty much needs to be a subset of the Cipher API, as opposed to a new internal API.